api/repositories/app_repository_test.go

```diff
@@ -28,6 +31,263 @@ const (
 	CFAppStoppedState  = "STOPPED"
 )
 
+//counterfeiter:generate -o fake -fake-name ClientWithWatch sigs.k8s.io/controller-runtime/pkg/client.WithWatch
+var _ = Describe("AppRepository Units", func() {
+	var (
+		testCtx                context.Context
+		appRepo                *AppRepo
+		fakeUserClient         fake.ClientWithWatch
+		fakeFactory            fake.UserK8sClientFactory
+		fakeNamespaceRetriever fake.NamespaceRetriever
+		fakeNsPerms            fake.NamespacePermissions
+	)
+
+	BeforeEach(func() {
+		testCtx = context.Background()
+		fakeUserClient = fake.ClientWithWatch{}
+		fakeFactory = fake.UserK8sClientFactory{}
+		fakeFactory.BuildClientReturns(&fakeUserClient, nil)
+		fakeNsPerms = fake.NamespacePermissions{}
+		fakeNamespaceRetriever = fake.NamespaceRetriever{}
+		appRepo = NewAppRepo(&fakeNamespaceRetriever, &fakeFactory, &fakeNsPerms)
+	})
+
+	Describe("GetApp", func() {
+		var retErr error
+		JustBeforeEach(func() {
+			_, retErr = appRepo.GetApp(testCtx, authInfo, "some-app-name")
+		})
+
+		When("the client factory reports an error", func() {
```

Comment:
I would aim to cover as much as possible with the unit tests. So as well as covering the error paths, e.g. for GetApp, we should also:
- check that the namespaceRetriever is invoked with the correct args
- check the expected authInfo is passed to the userClientFactory
- check that the expected namespace and name are requested in the userClient.Get

We saw a case recently where a couple of string arguments were passed in the wrong order, so as well as covering the untested error paths, these tests could guarantee calling out correctly.

Reply:
I forgot to comment here that I pushed a change to add the arg checks you mentioned. I chose to skip the assertions against the userClient, though, since the test-env tests already verify that the correct data is being fetched. The test-env tests will also survive a refactor (e.g adding an indexed field to a resource), where the units will require changes.